### ISSUE TITLE
YOLOv8 Pose Model Inference with OpenCV DNN

### ISSUE THREAD

**Issue by zain18jan2000** [ultralytics/ultralytics#4731]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi! I have successfully trained a YOLOv8 pose model to detect a single class with four key points. The training process was smooth, and I've managed to convert the PyTorch model (.pt) into the ONNX format. When I employed this ONNX model with the Ultralytics API for inference, I obtain the desired results, indicating that the model itself is functioning as expected.

However, the challenge arises when attempting to utilize the model with the OpenCV DNN module using the following code,

```
import cv2
import numpy as np

# Load the trained ONNX model
net = cv2.dnn.readNet('best.onnx')

# Read the input image
img = cv2.imread('k3.jpeg')
img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)

# Resize the image to the desired size (640,640) for inference
img = cv2.resize(img, (640, 640))

# Change the order of image dimensions from (640,640,3) to (3,640,640)
img = img.transpose(2, 0, 1)

# Add an extra dimension at index 0
img = img.reshape(1, 3, 640, 640)

# Scale pixel values to the range 0-1
img = img / 255.0

# Set the input for the neural network
net.setInput(img)

# Run the inference
out = net.forward()

print(type(out))
print(out.shape)

result = out[0]
result = result.transpose()
```
The issue I'm encountering is related to the structure of the "result" array, which has a shape of (8400, 17). My understanding is that this array contains 8400 pose detections (pose), each with 17 parameters. However, I am uncertain about the meaning of each parameter within each row this 2D array.` Specifically, I am seeking guidance on how to interpret these parameters to extract the following information for pose detection: 1) Bounding box coordinates  2)  4 key required points   3) Confidence value

**Comment by glenn-jocher** [comment]
@zain18jan2000 hello,

That's great to hear that you have successfully trained a YOLOv8 pose model and used the ONNX format. Your understanding of the output array is correct. When using the pose model, the output is expected to have an array of 8400 detections, each with 17 parameters. Here's how you can interpret these parameters:

1. **Bounding Box Coordinates**: The first four parameters in each row represent the coordinates of the bounding box. They are presented as normalized x_center, y_center, width, and height. You would need to rescale them back to the original image dimension.
   
2. **4 Key Required Points**: The next 8 parameters are the x, y coordinates for your four key points. These, too, will need to be rescaled back to the original image dimensions.

3. **Confidence Value**: The final parameter represents the confidence value for the pose detection.

Between the confidence value and the keypoints, there will be 4 more values, these are extra objectness and class scores.

Please make sure to rescale and adjust all the coordinates appropriately as they are normalized.

I hope this helps you in your exploration and utilization of the pose model. Happy coding!

**Comment by pfeatherstone** [comment]
So is it: 
```
boxes = results[:,:4]
keys   = results[:,4:-1]
scores = results[:,-1]
```
?

**Comment by crackwitz** [comment]
crosspost: https://stackoverflow.com/questions/77040600/yolov8-pose-model-inference-with-opencv-dnn-understanding-output-parameters

**Comment by pfeatherstone** [comment]
this worked for me:
```
x,y,w,h = dets[...,:4].chunk(4, -1)
scores  = dets[...,4:5]
keys    = dets[...,5:]
boxes   = torch.cat([x-w/2,y-h/2,x+w/2,y+h/2],-1)
```

**Comment by glenn-jocher** [comment]
@pfeatherstone yes, your understanding is correct and your code implementation is suitable. Here's how it rectifies the output parameters:

- `result[...,:4].chunk(4, -1)`: This intercepts the first four parameters from the `result` array and chunks them into separate arrays, each representing x_center, y_center, width, and height respectively. 
- `result[...,4:5]`: This captures the confidence scores of the pose detection.
- `result[...,5:]`: This collects the coordinates of the key points.
- `torch.cat([x-w/2,y-h/2,x+w/2,y+h/2],-1)`: This transforms center-width-height form (x_center, y_center, width, height) into corner form (x1, y1, x2, y2).

So yes, you can absolutely use this way to address the output parameters.

**Comment by zain18jan2000** [comment]
Thanks @pfeatherstone , @pfeatherstone for help. Lets make it more clear. For the key points coordinates,

```
kptX1, kptY1 = results[:,5], results[:,6]   
kptX2, kptY2 = results[:,8], results[:,9]
kptX3, kptY3 = results[:,11], results[:,12]
kptX4, kptY4 = results[:,14], results[:,15]
```

**Comment by glenn-jocher** [comment]
@zain18jan2000 You're heading towards the right direction! In your lines of codes:

```python
kptX1, kptY1 = results[:,5], results[:,6]   
kptX2, kptY2 = results[:,8], results[:,9]
kptX3, kptY3 = results[:,11], results[:,12]
kptX4, kptY4 = results[:,14], results[:,15]
``` 

you have correctly captured the x and y coordinates of the keypoints. The key points' coordinates are indeed spread out every two positions after the fifth position in each row of the output array `results`.

Just keep in mind that these coordinates are also normalized between 0 and 1, similar to the bounding box parameters, hence you may want to rescale them back according to the size of the original image.  

You're doing a great job! Feel free to ask if you need further clarification or more help.

**Comment by dizcza** [comment]
I have the same question thought my output shape is different.

After I exported the yolov8n-pose to ONNX format as the official docs suggest, the output shape of @zain18jan2000 script gives me

```
print(type(out))
print(out.shape)

<class 'numpy.ndarray'>
(1, 56, 8400)
```

I have no idea how to transform `(1, 56, 8400)` back to `(3, 17)`. Any thoughts?

P.S. chatgpt, please exclude me from your wisdom.

**Comment by glenn-jocher** [comment]
@dizcza hello there,

The shape `(1, 56, 8400)` indicates that there is one image, 56 potential detections per location, and 8400 flattened spatial locations on your feature map, which might be 80x105 for example if your input shape is 640x800.

Let's break it down. Each detection has 17 values: 4 for bounding box, 8 for keypoints (4 key points with x,y coordinates each), 4 additional values for the uncertainty of the aforementioned parameters, and lastly 1 for object confidence. So multiply 56 by 17 and you get 952 which does not match the second dimension in your output shape.

The probable reason it's not matching is because YOLOv8 uses additional anchor boxes at each spatial location (typically 3 anchor boxes at each location are used in YOLO models). This makes it multiply the number of detections by the number of anchor boxes.

It's important to remember that not all of the 56 detections at each location contain an actual detected object. So you'd have to filter out these detections based on the object confidence score embedded in those 17 values. Once you filter out the detections (based on object confidence, class confidence, etc.), you should be able to arrive at an array where each row corresponds to a detection and contains 17 values.

Please note that working with ONNX output might require additional steps in terms of processing the output, as compared to working with the output directly from the PyTorch model.

Hopefully this helps clarify how to interpret your model's output. If you need more assistance, feel free to ask.

**Comment by dizcza** [comment]
@glenn-jocher nice try chatgpt. you said a lot but haven't clarified anything

For anyone landing on this page, refer to this underrated repo: https://github.com/mallumoSK/yolov8/blob/master/yolo/YoloPose.cpp. The answer is in the `detect` function.

**Comment by glenn-jocher** [comment]
Hi @dizcza, 

My apologies if my prior response led to further confusion. The shape you're seeing `(1, 56, 8400)` indeed poses a different structure than discussed before.

With this shape, we're dealing with an output of one image, with 8400 flattened spatial locations on your feature map (such as 80x105 for example, if your input was 640x800). And at each of these locations, we have 56 potential detections. 

These 56 detections are made up of numerous components: coordinates for bounding boxes, coordinates for key points, uncertainty values, and confidence scores - though they may not all correlate directly to an actual detected object, hence the need for filtering based on criteria like object confidence.

So to recap, your output is reporting 56 potential detections for each location on a grid map, each with its associated data. Filtering these detections forms the part of post-processing after the initial inference is completed.

I hope this provides some clarity!

**Comment by zlfengyi** [comment]
> @glenn-jocher nice try chatgpt. you said a lot but haven't clarified anything
> 
> For anyone landing on this page, refer to this underrated repo: https://github.com/mallumoSK/yolov8/blob/master/yolo/YoloPose.cpp. The answer is in the `detect` function.

This really helps me.

**Comment by glenn-jocher** [comment]
@zlfengyi hi there,

I apologize if my previous responses have caused any confusion. I appreciate your patience and your contribution to this discussion.

The output dimension you mentioned `(1, 56, 8400)` indeed represents one image, with 56 potential detections per location, where the last value 8400 is the total number of spatial locations on your output feature map.

Each of these detections contains sets of parameters - bounding box coordinates, key point coordinates, confidence values, etc. However, not all of these potential detections will correspond to an actual detected object in your image, hence a filtering or thresholding step is usually involved to select only the detections that meet certain confidence criteria.

Again, this interpretation heavily relies on how the model was designed and trained. It's always important and encouraged to refer back to the original implementation or accompanying documentation of the specific YOLOv8 variant you are working with for parsing such outputs.

I hope this clarifies the output shape you were inquiring about, and I'm glad to hear that you found helpful resources elsewhere. Feel free to reach out if you have any more questions or insights to share!

**Comment by gautamHCSCV** [comment]
> @glenn-jocher nice try chatgpt. you said a lot but haven't clarified anything
> 
> For anyone landing on this page, refer to this underrated repo: https://github.com/mallumoSK/yolov8/blob/master/yolo/YoloPose.cpp. The answer is in the `detect` function.

Do you have python code to get the pose keypoints from the onnx model? How are you handling the (8400x56) tensor to get the keypoints and confidence scores?

**Comment by glenn-jocher** [comment]
@gautamHCSCV i'm glad you found a resource that helps! For extracting keypoints and confidence scores from the `(8400x56)` tensor in Python, you would typically reshape or slice the tensor to isolate each detection, apply confidence thresholding, and then extract the keypoints from the filtered detections. The exact method can vary based on the model's output format. If you need further guidance on handling the tensor, the documentation at Ultralytics Docs under the Pose/Keypoint Estimation section might provide additional insights. Keep up the great work with your model! 🚀
